Mathlib/Tactic/ToExpr.lean

```diff
@@ -0,0 +1,81 @@
+/-
+Copyright (c) 2023 Kyle Miller. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kyle Miller
+-/
+import Mathlib.Tactic.DeriveToExpr
+
+/-! # `ToExpr` instances for Mathlib
+
+This module should be imported by any module that intends to define `ToExpr` instances.
+It provides necessary dependencies (the `Lean.ToLevel` class) and it also overrides the instances
+that come from core Lean 4 that do not handle universe polymorphism.
+(See the module `Lean.ToExpr` for the instances that are overridden.)
+
+In addition, we provide some additional `ToExpr` instances for core definitions.
+-/
+
+section override
+namespace Lean
+
+attribute [-instance] Lean.instToExprOption
+
+deriving instance ToExpr for Option
+
+attribute [-instance] Lean.instToExprList
+
+deriving instance ToExpr for List
+
+attribute [-instance] Lean.instToExprArray
+
+instance {α : Type u} [ToExpr α] [ToLevel.{u}] : ToExpr (Array α) :=
+  let type := toTypeExpr α
+  { toExpr     := fun as => mkApp2 (mkConst ``List.toArray [toLevel.{u}]) type (toExpr as.toList)
+    toTypeExpr := mkApp (mkConst ``Array [toLevel.{u}]) type }
+
+attribute [-instance] Lean.instToExprProd
+
+deriving instance ToExpr for Prod
+
+end Lean
+end override
+
+namespace Mathlib
+open Lean
+
+deriving instance ToExpr for Int
+
+deriving instance ToExpr for ULift
+
+/-- Hand-written instance since `PUnit` is a `Sort` rather than a `Type`. -/
+instance [ToLevel.{u}] : ToExpr PUnit.{u+1} where
+  toExpr _ := mkConst ``PUnit.unit [toLevel.{u+1}]
+  toTypeExpr := mkConst ``PUnit [toLevel.{u+1}]
+
+open DataValue in
+private def toExprMData (md : MData) : Expr := Id.run do
+  let mut e := mkConst ``MData.empty
+  for (k, v) in md do
+    let k := toExpr k
+    e := match v with
+          | ofString v => mkApp3 (mkConst ``KVMap.setString) e k (mkStrLit v)
+          | ofBool v   => mkApp3 (mkConst ``KVMap.setBool) e k (toExpr v)
+          | ofName v   => mkApp3 (mkConst ``KVMap.setName) e k (toExpr v)
+          | ofNat v    => mkApp3 (mkConst ``KVMap.setNat) e k (mkNatLit v)
+          | ofInt v    => mkApp3 (mkConst ``KVMap.setInt) e k (toExpr v)
+          | ofSyntax _ => panic! "toExpr for Syntax is not implemented in toExprMData"
+  return e
```

Comment:
@gebner I just lifted this from [QQ/Reflect](https://github.com/gebner/quote4/blob/8fd8422dfe2cd5d061754944ca345ffa6061781b/Qq/Reflect.lean), but there's now an `ofSyntax` case to deal with. I put in a panic for now -- do you think that's OK temporarily, or should I delete the `ToExpr Expr` instance if I don't plan on finishing it myself?

Reply:
I think it's okay to leave it as a TODO.  I'm not sure if I prefer a panic (which we prohibit in mathlib) or simply `e` (i.e. ignore the value) as a placeholder.

Reply:
Ok, I've added a TODO and chose a third option for now: insert a string into the table with the error. I don't really know what uses `Syntax` in `MData`, so I can't tell if this is a bad idea, but we can easily switch it to dropping the value like you suggested.

Reply:
Actually, it turned out it just needed a handful more `deriving instance` lines, so the TODO is done.